debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorTask.java

```diff
@@ -0,0 +1,146 @@
+    private static final String CONTEXT_NAME = "oracle-connector-task";
+
+    private static enum State {
+        RUNNING, STOPPED;
+    }
+
+    private volatile OracleTaskContext taskContext;
+    private volatile ChangeEventQueue<Object> queue;
+    private volatile OracleConnection jdbcConnection;
+    private volatile ChangeEventSourceCoordinator coordinator;
+    private volatile ErrorHandler errorHandler;
+    private final AtomicReference<State> state = new AtomicReference<State>(State.STOPPED);
+
+    @Override
+    public String version() {
+        return Module.version();
+    }
+
+    @Override
+    public void start(Configuration config) {
+        if (!state.compareAndSet(State.STOPPED, State.RUNNING)) {
+            LOGGER.info("Connector has already been started");
+            return;
+        }
+
+        OracleConnectorConfig connectorConfig = new OracleConnectorConfig(config);
+        taskContext = new OracleTaskContext(connectorConfig);
+
+        Clock clock = Clock.system();
+
+        // Set up the task record queue ...
+        this.queue = new ChangeEventQueue.Builder<Object>()
+                .pollInterval(connectorConfig.getPollInterval())
+                .maxBatchSize(connectorConfig.getMaxBatchSize())
+                .maxQueueSize(connectorConfig.getMaxQueueSize())
+                .loggingContextSupplier(() -> taskContext.configureLoggingContext(CONTEXT_NAME))
+                .build();
+
+        errorHandler = new ErrorHandler(OracleConnector.class, connectorConfig.getLogicalName(), queue, this::cleanupResources);
+        TopicSelector topicSelector = OracleTopicSelector.defaultSelector(connectorConfig.getLogicalName());
+
+        Configuration jdbcConfig = config.subset("database.", true);
+
+        jdbcConnection = new OracleConnection(jdbcConfig, new OracleConnectionFactory());
+        SchemaNameAdjuster schemaNameAdjuster = SchemaNameAdjuster.create(LOGGER);
+        OracleDatabaseSchema schema = new OracleDatabaseSchema(schemaNameAdjuster, topicSelector);
+
+        EventDispatcher dispatcher = new EventDispatcher(errorHandler, topicSelector, schema, queue);
+
+        coordinator = new ChangeEventSourceCoordinator(
+                errorHandler,
+                OracleConnector.class,
+                connectorConfig.getLogicalName(),
+                new OracleChangeEventSourceFactory(connectorConfig, jdbcConnection, errorHandler, dispatcher, clock, schema)
+        );
+
+        coordinator.start();
+    }
+
+    @Override
+    public List<SourceRecord> poll() throws InterruptedException {
+        // TODO
+        List records = queue.poll();
+
+        return ((List<DataChangeEvent>)records).stream()
+            .map(DataChangeEvent::getRecord)
+            .collect(Collectors.toList());
+    }
+
+    @Override
+    public void stop() {
+        cleanupResources();
+    }
+
+    private void cleanupResources() {
+        if (!state.compareAndSet(State.RUNNING, State.STOPPED)) {
```

Comment:
I think this might be called by two threads in parallel. If an error happens during termination (after a `stop()` was called) of change event source then `errorHandler` is invoked asynchornously that will call `cleanupResources()` again

Reply:
Ok, I'll take a look at that one.